### DISCUSSION TITLE
Scoping Custom Events Only to the Connected Client

### DISCUSSION THREAD

**Discussion by ProbablyBrian** [zauberzeug/nicegui#2477] [Q&A] [answered]
### Question

I'm trying to create a table with an accompanying details panel on the side that displays more information about a row.  I've defined a custom body slot for a column whose value should be a link that, upon the user clicking it, refreshes the information in the details panel.

I've almost got it working with one caveat: the details panel refreshes for all connected users when anyone viewing the page clicks a link.  I'd like to "scope" this event to only refresh the details panel for the user performing the triggering event, although I'm not sure how to accomplish this.

To reproduce, first have a look at the following code example:

```python
from nicegui import ui

columns = [
    {
        "name": "name",
        "label": "Name",
        "field": "name",
        "align": "left",
        "classes": "nicegui-link",
    },
    {"name": "link", "label": "Link", "field": "link", "align": "left"},
]
rows = [
    {"name": "Google", "link": "https://google.com"},
    {"name": "Facebook", "link": "https://facebook.com"},
    {"name": "Twitter", "link": "https://twitter.com"},
]


@ui.page("/")
async def homepage():
    with ui.row().classes("w-1/2"):
        table = await build_table()
        await details_panel(row=None)


async def build_table():
    table = ui.table(rows=rows, columns=columns, row_key="name")
    table.add_slot(
        "body-cell-name",
        """
        <q-td :props="props">
            <a href="javascript:;" @click="$parent.$emit('highlight', props)">{{ props.value }}</a>
        </q-td>
    """,
    )
    table.on(
        "highlight",
        lambda e: details_panel.refresh(row=e.args["row"]),
        args=["row"],
    )


@ui.refreshable
async def details_panel(row: dict[str, str] = None):
    with ui.column().classes("gap-0"):
        ui.label("Item Details").classes("text-2xl")
        if not row:
            ui.label("No selection").classes("text-gray-400")
            return

        for key, value in row.items():
            ui.label(key).classes("text-gray-400 pt-4")
            ui.label(value)


ui.run()
```

Running the above, the behavior is as expected.  However, when i connect another client to the page and click a link, all of the connected clients refresh their details panel with the row corresponding to the link that was clicked.

What can i do in order to scope the "highlight" event to only the client that actually clicked it?

Note:  In my actual project, I'm already making use of selection="multiple" on the table for selecting rows to perform bulk actions (delete, update, etc) from a drop-down menu.  I realize I could also use the rowClick event, although I'd prefer to have an element that's clearly identified as interactive for improved accessibility for screen-reader software.

**Comment by falkoschindler** [comment, marked as answer]
Hi @ProbablyBrian,

Defining the refreshable function inside the page makes it individual for the current client:
```py
columns = [
    {"name": "name", "label": "Name", "field": "name", "align": "left", "classes": "nicegui-link"},
    {"name": "link", "label": "Link", "field": "link", "align": "left"},
]
rows = [
    {"name": "Google", "link": "https://google.com"},
    {"name": "Facebook", "link": "https://facebook.com"},
    {"name": "Twitter", "link": "https://twitter.com"},
]

@ui.page("/")
def homepage():
    @ui.refreshable
    def details_panel(row: dict[str, str] = {}):
        with ui.column().classes("gap-0"):
            ui.label("Item Details").classes("text-2xl")
            if not row:
                ui.label("No selection").classes("text-gray-400")
                return
            for key, value in row.items():
                ui.label(key).classes("text-gray-400 pt-4")
                ui.label(value)

    with ui.row().classes("w-1/2"):
        table = ui.table(rows=rows, columns=columns, row_key="name")
        table.add_slot("body-cell-name", """
            <q-td :props="props">
                <a href="javascript:;" @click="$parent.$emit('highlight', props)">{{ props.value }}</a>
            </q-td>
        """)
        table.on("highlight", lambda e: details_panel.refresh(row=e.args["row"]), args=["row"])
        details_panel(row=None)
```

Does this solve your problem?

**Reply by ProbablyBrian** [reply]
Thanks for the quick reply @falkoschindler 

Yes, that does work! I incorrectly expected it to work similarly to other functions that create nicegui objects within the function body.  But, I completely overlooked the fact it's a decorator and that the ui.refreshable isn't itself being defined in a function but in the global scope.  Kind of obvious now that I think about it but in the moment, this wisdom eluded me.

I did notice that you collapsed everything into one function.  Is there any harm or problems with doing something like the below?

```python
...

@ui.page("/")
async def homepage():
    details_panel = ui.refreshable(build_details_panel)
    with ui.row().classes("w-1/2"):
        table = await build_table(details_panel=details_panel)
        await details_panel(row=None)

async def build_table(details_panel: ui.refreshable):
    ...

async def build_details_panel(row: dict[str, str] | None = None):
    ...
```

It produces the same end-results since the ui.refreshable element is still defined in the page function's scope.  It isn't an issue here, but being able to not nest/indent so heavily in large nicegui page functions has proved useful for keeping things organized.  If there's a more optimal and/or recommended way of breaking up larger page functions, I'd love to learn them to use the package more effectively.

P.S. Good catch on the type annotations being off for the details_panel function :)

**Reply by falkoschindler** [reply]
Yes, that looks like a good solution. At least I couldn't come up with something simpler. You could also pass `details_panel.refresh` in form of a `Callable` to `build_table`, since that's all this function needs. But that's just a minor detail.

P.S.: I like to have a mypy type checker running in VSCode at all times. This helps to catch bugs very early. 😉
